src/node_buffer.cc

```diff
@@ -855,85 +856,147 @@ void Compare(const FunctionCallbackInfo<Value> &args) {
 }
 
 
-int32_t IndexOf(const char* haystack,
-                size_t h_length,
-                const char* needle,
-                size_t n_length) {
-  CHECK_GE(h_length, n_length);
-  // TODO(trevnorris): Implement Boyer-Moore string search algorithm.
-  for (size_t i = 0; i < h_length - n_length + 1; i++) {
-    if (haystack[i] == needle[0]) {
-      if (memcmp(haystack + i, needle, n_length) == 0)
-        return i;
-    }
-  }
-  return -1;
-}
-
-
 void IndexOfString(const FunctionCallbackInfo<Value>& args) {
   ASSERT(args[1]->IsString());
   ASSERT(args[2]->IsNumber());
+  ASSERT(args[3]->IsString());
+
+  enum encoding enc = ParseEncoding(args.GetIsolate(),
+                                    args[3].As<String>(),
+                                    UTF8);
 
   THROW_AND_RETURN_UNLESS_BUFFER(Environment::GetCurrent(args), args[0]);
   SPREAD_ARG(args[0], ts_obj);
 
-  node::Utf8Value str(args.GetIsolate(), args[1]);
-  int32_t offset_i32 = args[2]->Int32Value();
-  uint32_t offset;
+  Local<String> needle = args[1]->ToString();
+  const char* haystack = ts_obj_data;
+  int32_t haystack_length = ts_obj_length;
 
-  if (offset_i32 < 0) {
-    if (offset_i32 + static_cast<int32_t>(ts_obj_length) < 0)
+  if (enc == UCS2) {
+    haystack_length /= 2;
```

Comment:
I'm still confused about this. No matter what the needle's encoding, it can be turned into a byte sequence. So there should be no need to adjust the length of the haystack. It should be basically the same as passing:
```js
buffer.indexOf(new Buffer(<string>, <encoding>));
```